autoprecompiles/src/equivalence_classes.rs

```diff
@@ -2,74 +2,164 @@ use std::collections::{BTreeSet, HashMap};
     /// Singleton equivalence classes are omitted from the result.
     pub fn intersect(partitions: &[Self]) -> Self {
-        // For each partition, build a map: Id -> class_index
-        let class_ids: Vec<HashMap<T, usize>> = partitions
+        let Some(first) = partitions.first() else {
+            return Self {
+                class_of: HashMap::new(),
+                num_classes: 0,
+            };
+        };
+
+        // Pairwise intersection: fold over partitions, intersecting two at a time.
+        // This is more efficient than building Vec<usize> signatures because:
+        // 1. We only hash (usize, usize) tuples instead of Vec<usize>
+        // 2. The result shrinks after each intersection, making later steps faster
+        partitions[1..]
             .iter()
-            .map(|partition| {
-                partition
-                    .iter()
-                    .enumerate()
-                    .flat_map(|(class_idx, class)| class.iter().map(move |&id| (id, class_idx)))
-                    .collect()
-            })
-            .collect();
+            .fold(first.clone(), |acc, p| Self::intersect_two(&acc, p))
+    }
 
-        // Iterate over all elements in the universe
-        partitions
+    /// Intersects two partitions.
+    fn intersect_two(a: &Self, b: &Self) -> Self {
+        // Group elements by (class_in_a, class_in_b)
+        // Elements with the same pair end up in the same result class
+        a.class_of
             .iter()
-            .flat_map(|partition| partition.iter())
-            .flat_map(|class| class.iter())
-            .unique()
-            .filter_map(|id| {
-                // Build the signature of the element: the list of class indices it belongs to
-                // (one index per partition)
-                class_ids
-                    .iter()
-                    .map(|m| m.get(id).cloned())
-                    // If an element did not appear in any one of the partitions, it is
-                    // a singleton and we skip it.
-                    .collect::<Option<Vec<usize>>>()
-                    .map(|signature| (signature, id))
+            // Note that if an element is not in a or b, it is a singleton and will also
+            // not be in the intersection.
+            .filter_map(|(&elem, &class_a)| {
+                b.class_of
+                    .get(&elem)
+                    .map(|&class_b| ((class_a, class_b), elem))
             })
-            // Group elements by their signatures
             .into_group_map()
             .into_values()
-            // Convert to set
-            .map(|ids| ids.into_iter().copied().collect())
+            .collect()
+    }
+}
+
+/// Threshold below which we use sequential intersection.
+const PARALLEL_THRESHOLD: usize = 16;
+
+/// Number of partitions to combine in each chunk before parallelizing.
+const CHUNK_SIZE: usize = 64;
+
+impl<T: Eq + Hash + Copy + Send + Sync> Partition<T> {
```

Comment:
Do we need these complicated bounds? Could be get away with `FieldElement` ?

Reply:
Oh right it's not the value. If it was only `BlockCell` I would say we should hardcode but the tests use integers so it's fine as is.